File: frontend/src/utils/fiscalCalc.js
// Calcolo di ritenuta d'acconto e contributo ENASARCO su una provvigione
// lorda — unica fonte della formula lato frontend, usata sia dal
// calcolatore "isolato" del blog (RitenutaEnasarcoCalculator.jsx, un
// numero inserito a mano) sia dal riepilogo fiscale reale nella pagina
// Provvigioni (che applica la stessa formula ai totali già calcolati dal
// sistema). Stessa logica, stesse aliquote, del backend
// (services/fiscal_calc.py) — le due copie esistono perché il blog deve
// restare 100% client-side (nessuna chiamata API da un articolo pubblico),
// mentre il riepilogo reale in app lavora su dati già caricati lato
// client: tenerle esplicitamente allineate è più semplice che introdurre
// una chiamata di rete solo per un calcolo aritmetico.
//
// Riferimenti normativi (verificati per l'articolo del blog "ritenuta
// d'acconto e contributi ENASARCO"):
// - Ritenuta d'acconto: art. 25-bis DPR 600/1973, aliquota nominale 23%,
//   su base 50% (ordinaria) o 20% (ridotta, richiede dichiarazione
//   formale al mandante entro il 31/12 dell'anno precedente).
// - Esenzione totale dalla ritenuta per il regime forfettario: L.
//   190/2014, comma 67.
// - Contributo ENASARCO 2026: 17% totale, 8,5% a carico dell'agente,
//   indipendente dal regime fiscale (forfettario incluso).
export const RITENUTA_ALIQUOTA = 0.23;
export const ENASARCO_QUOTA_AGENTE = 0.085;

// Massimali/minimali contributivi ENASARCO 2026, un tetto per ciascun
// rapporto di agenzia (non complessivo su più mandanti) — verificati per
// l'articolo del blog "ENASARCO: i nuovi minimali e massimali provvigionali
// 2026". Unica fonte: prima esisteva solo una copia locale dentro
// RitenutaEnasarcoCalculator.jsx, che non poteva essere riusata dal
// riepilogo fiscale reale dell'app (vedi computeEnasarcoConMassimale sotto).
export const ENASARCO_SOGLIE = {
  plurimandatario: { massimale: 30478, minimale: 515 },
  monomandatario: { massimale: 45717, minimale: 1026 },
};

// Arrotonda a 2 decimali con Math.round (non un semplice toFixed, che
// tronca): tiene i valori restituiti allineati a come li arrotonda il
// gemello Python (services/fiscal_calc.py usa round(x, 2) su ogni campo) —
// senza questo, un confronto diretto tra i due (es. un futuro endpoint che
// espone il calcolo Python allo stesso frontend) potrebbe mostrare un
// centesimo di differenza sullo stesso importo. I nomi dei campi restano
// invece camelCase qui e snake_case lato Python, per convenzione idiomatica
// di ciascun linguaggio — un eventuale endpoint che esponga il calcolo
// Python dovrà comunque tradurli, non è un refuso.
export function round2(n) {
  return Math.round((n + Number.EPSILON) * 100) / 100;
}

// Contributo ENASARCO su un insieme di importi attribuiti a mandanti (già
// filtrati da chi chiama per status/anno rilevanti), rispettando il
// massimale annuo PER SINGOLO MANDANTE — un plurimandatario con 3 mandanti
// ha 3 tetti separati da 30.478€, non un unico tetto da dividere tra loro
// (stesso punto verificato per l'articolo minimali/massimali). Importi
// senza un mandante noto (es. una provvigione manuale non attribuita a
// nessuno) non hanno un tetto applicabile: trattati con l'aliquota piena,
// senza cap — stesso comportamento di prima dell'introduzione di questa
// funzione, quando il massimale non era considerato affatto.
//
// items: [{ mandanteId: string|null, amount: number }]
// mandantiById: { [id]: { esclusiva?: boolean } }
export function computeEnasarcoConMassimale(items, mandantiById) {
  const perMandante = new Map();
  let senzaMandante = 0;
  for (const { mandanteId, amount } of items || []) {
    const a = Math.max(0, amount || 0);
    if (!mandanteId) {
      senzaMandante += a;
      continue;
    }
    perMandante.set(mandanteId, (perMandante.get(mandanteId) || 0) + a);
  }

  let contributo = senzaMandante * ENASARCO_QUOTA_AGENTE;
  const dettagliPerMandante = [];
  for (const [mandanteId, cumulato] of perMandante) {
    const esclusiva = !!mandantiById?.[mandanteId]?.esclusiva;
    const soglie = esclusiva ? ENASARCO_SOGLIE.monomandatario : ENASARCO_SOGLIE.plurimandatario;
    const imponibile = Math.min(cumulato, soglie.massimale);
    contributo += imponibile * ENASARCO_QUOTA_AGENTE;
    dettagliPerMandante.push({
      mandanteId,
      cumulato,
      massimale: soglie.massimale,
      minimale: soglie.minimale,
      superaMassimale: cumulato > soglie.massimale,
      sottoMinimale: cumulato < soglie.minimale,
    });
  }

  return { contributo: round2(contributo), dettagliPerMandante };
}

export function computeFiscalBreakdown(lordo, regimeFiscale, baseRitenuta) {
  const lordoSafe = Math.max(0, lordo || 0);
  const baseImponibile =
    regimeFiscale === "ordinario" ? (baseRitenuta === "50" ? 0.5 : 0.2) : 0;
  const ritenuta = lordoSafe * baseImponibile * RITENUTA_ALIQUOTA;
  const enasarco = lordoSafe * ENASARCO_QUOTA_AGENTE;
  return {
    lordo: round2(lordoSafe),
    ritenutaAcconto: round2(ritenuta),
    contributoEnasarco: round2(enasarco),
    netto: round2(lordoSafe - ritenuta - enasarco),
  };
}

export function formatEuro(n) {
  return (n || 0).toLocaleString("it-IT", { style: "currency", currency: "EUR" });
}

// Converte un numero digitato in formato italiano (punto = separatore
// delle migliaia, virgola = decimale) in un float JS. Senza questo,
// parseFloat("15.000".replace(",", ".")) restituisce 15 invece di 15000,
// perché "." è già un punto decimale valido in JS — un refuso di tre
// ordini di grandezza, silenzioso, se l'utente digita l'importo con il
// formato italiano standard (es. "15.000" per quindicimila euro).
export function parseItalianNumber(str) {
  const normalized = String(str ?? "").replace(/\./g, "").replace(",", ".");
  return parseFloat(normalized) || 0;
}

File: backend/services/fiscal_calc.py
"""Calcolo di ritenuta d'acconto e contributo ENASARCO su una provvigione
lorda reale. Stessa identica logica (e stesse aliquote) del calcolatore
interattivo pubblicato nell'articolo del blog "ritenuta d'acconto e
contributi ENASARCO" (frontend/src/components/RitenutaEnasarcoCalculator.jsx
/ frontend/src/utils/fiscalCalc.js) — tenuta qui come funzione pura e
riusabile perché, a differenza di quel calcolatore (un numero inserito a
mano, isolato), qui si applica a dati reali già presenti nel sistema
(commission_service), quindi deve restare importabile da più punti
(endpoint di riepilogo, dashboard, export) senza duplicare la formula.

Riferimenti normativi (verificati per l'articolo del blog):
- Ritenuta d'acconto: art. 25-bis DPR 600/1973, aliquota nominale 23%,
  applicata su base 50% (ordinaria) o 20% (ridotta, richiede dichiarazione
  formale al mandante entro il 31/12 dell'anno precedente).
- Esenzione totale dalla ritenuta per il regime forfettario: L. 190/2014,
  comma 67.
- Contributo ENASARCO 2026: 17% totale, 8,5% a carico dell'agente,
  indipendente dal regime fiscale (forfettario incluso), calcolato su un
  imponibile con massimale/minimale annuo PER MANDANTE (non complessivo):
  30.478€/515€ plurimandatario, 45.717€/1.026€ monomandatario/esclusiva —
  vedi l'articolo blog sui minimali/massimali 2026 e il gemello JS
  (frontend/src/utils/fiscalCalc.js, ENASARCO_SOGLIE/
  computeEnasarcoConMassimale).
"""

from typing import Literal, TypedDict

RITENUTA_ALIQUOTA = 0.23
ENASARCO_QUOTA_AGENTE = 0.085

ENASARCO_SOGLIE = {
    "plurimandatario": {"massimale": 30478.0, "minimale": 515.0},
    "monomandatario": {"massimale": 45717.0, "minimale": 1026.0},
}


class EnasarcoConMassimale(TypedDict):
    imponibile: float
    contributo_enasarco: float
    supera_massimale: bool
    sotto_minimale: bool


def compute_enasarco_con_massimale(
    cumulato_prima: float, lordo_periodo: float, esclusiva: bool
) -> EnasarcoConMassimale:
    """Calcola il contributo ENASARCO su `lordo_periodo` tenendo conto di
    quanto già maturato con lo stesso mandante nell'anno prima di questo
    periodo (`cumulato_prima`): solo la parte di `lordo_periodo` che rientra
    ancora nel massimale annuo residuo è imponibile — il resto è esente. Vedi
    RitenutaEnasarcoCalculator.jsx (stesso concetto "cumulativo prima" nel
    calcolatore interattivo) per la logica gemella lato utente."""
    soglie = ENASARCO_SOGLIE["monomandatario" if esclusiva else "plurimandatario"]
    cumulato_prima = max(0.0, cumulato_prima)
    lordo_periodo = max(0.0, lordo_periodo)
    residuo = max(0.0, soglie["massimale"] - cumulato_prima)
    imponibile = min(lordo_periodo, residuo)
    cumulato_totale = cumulato_prima + lordo_periodo
    return {
        "imponibile": round(imponibile, 2),
        "contributo_enasarco": round(imponibile * ENASARCO_QUOTA_AGENTE, 2),
        "supera_massimale": cumulato_totale > soglie["massimale"],
        "sotto_minimale": cumulato_totale < soglie["minimale"],
    }


# Nota per chi collegherà questa funzione a un endpoint reale (vedi sopra,
# "importabile da più punti"): il gemello JS (frontend/src/utils/fiscalCalc.js
# computeFiscalBreakdown) restituisce le stesse quattro grandezze ma in
# camelCase (ritenutaAcconto, contributoEnasarco), per convenzione idiomatica
# di ciascun linguaggio — non è un disallineamento accidentale, ma un
# endpoint che esponga questo dict come JSON dovrà comunque tradurre le
# chiavi se il consumer si aspetta il formato camelCase già in uso lato
# frontend. Entrambe le implementazioni arrotondano ora a 2 decimali con lo
# stesso criterio, per restare numericamente identiche a parità di input.
class FiscalBreakdown(TypedDict):
    lordo: float
    ritenuta_acconto: float
    contributo_enasarco: float
    netto: float


def compute_fiscal_breakdown(
    lordo: float,
    regime_fiscale: Literal["ordinario", "forfettario"],
    base_ritenuta: Literal["50", "20"],
) -> FiscalBreakdown:
    lordo = max(0.0, lordo)
    base_imponibile = 0.0
    if regime_fiscale == "ordinario":
        base_imponibile = 0.5 if base_ritenuta == "50" else 0.2
    ritenuta = lordo * base_imponibile * RITENUTA_ALIQUOTA
    enasarco = lordo * ENASARCO_QUOTA_AGENTE
    return {
        "lordo": round(lordo, 2),
        "ritenuta_acconto": round(ritenuta, 2),
        "contributo_enasarco": round(enasarco, 2),
        "netto": round(lordo - ritenuta - enasarco, 2),
    }
